src/autotester/server/utils/tests/file_management_test.py

```diff
@@ -0,0 +1,131 @@
+
+
+def test_clean_dir_name():
+    a = "markus/address"
+    b = a.replace("/", "_")
+    assert clean_dir_name(a) == b
+
+
+def test_random_tmpfile_name():
+    tmp_file = random_tmpfile_name()
+    dir_name = os.path.dirname(tmp_file)
+    base_name = os.path.basename(tmp_file)
+    assert os.path.exists(dir_name) and re.match(
+        r"[+\-]" "0" "[xX]" "0|" "([1-9A-Fa-f][1-9A-Fa-f]*)", base_name
+    )
+
+
+def test_recursive_iglob():
+    root_dir = tempfile.mkdtemp()
+    fd, file = tempfile.mkstemp(dir=root_dir)
+    sub_dir = tempfile.mkdtemp(dir=root_dir)
+    file_or_dir = list(recursive_iglob(root_dir))
+    dir = []
+    files = []
+    for i in file_or_dir:
+        dir.append(i[1]) if i[0] == "d" else files.append(i[1])
+    assert all([os.path.exists(d) for d in dir])
+    assert all([os.path.exists(f) for f in files])
+    assert sub_dir == dir[0]
+    assert file == files[0]
+    shutil.rmtree(root_dir)
+    os.close(fd)
+
+
+def test_copy_tree():
+    source_dir = tempfile.mkdtemp()
+    fd, source_file = tempfile.mkstemp(dir=source_dir)
+    dest_dir = tempfile.mkdtemp()
+    copy_tree(source_dir, dest_dir)
+    copied_file = os.path.join(dest_dir, source_file)
+    assert os.path.exists(copied_file)
+    os.close(fd)
+    shutil.rmtree(source_dir)
+    shutil.rmtree(dest_dir)
+
+
+def test_ignore_missing_dir_error():
+    dir = tempfile.mkdtemp()
+    shutil.rmtree(dir)
+    error = True
+    try:
+        shutil.rmtree(dir, onerror=ignore_missing_dir_error)
+    except Exception:
+        error = False
+    assert error
+
+
+def test_move_tree():
+    source_dir = tempfile.mkdtemp()
+    fd, source_file = tempfile.mkstemp(dir=source_dir)
+    dest_dir = tempfile.mkdtemp()
+    os.close(fd)
+    move_tree(source_dir, dest_dir)
+    file_name = os.path.basename(source_file)
+    moved_file = os.path.join(dest_dir, file_name)
+    assert os.path.exists(moved_file) and not os.path.exists(source_file)
+    shutil.rmtree(dest_dir)
+
+
+def test_fd_open():
+    dir = tempfile.mkdtemp()
+    dir_fd = os.open(dir, os.O_RDONLY)
+    file_fd, file = tempfile.mkstemp(dir=dir)
+    with fd_open(dir) as fdd:
+        same_dir = os.path.sameopenfile(fdd, dir_fd)
+    with fd_open(file) as fdf:
+        same_file = os.path.sameopenfile(fdf, file_fd)
+    assert same_dir and same_file
+    assert type(fdd) is int
+    assert type(fdf) is int
+    os.close(dir_fd)
+    os.close(file_fd)
+
+
+def test_fd_lock_p1():
+    temp_dir = tempfile.mkdtemp()
+    fd, temp_file = tempfile.mkstemp(dir=temp_dir)
+    access = True
+    with fd_open(temp_dir) as fd:
+        with fd_lock(fd):
+            try:
+                subprocess.run(["rm", {temp_file}])
+            except Exception:
+                access = False
+    assert not access
+
+
+def test_copy_test_script_files():
+    markus_address = "http://localhost:3000/csc108/en/main"
+    assignment_id = 1
+    tests_path = tempfile.mkdtemp()
```

Comment:
you use `tempfile.mkdtemp` and `tempfile.mkstemp` a lot. These are pretty low level functions and I think in a lot of cases you can use `tempfile.TemporaryDirectory` and `tempfile. NamedTemporaryFile` instead. That way it will clean up your tests as well because you won't have to worry about deleting the files/directories afterwards yourself